initialization/initialization.go

```diff
@@ -159,6 +159,21 @@ func withReferenceOracle(referenceOracle *oracle.WorkOracle) OptionFunc {
 	}
 }
 
+// Lazy initialized value of type T.
+type LazyInit[T any] struct {
+	init     func() (T, error)
+	initOnce sync.Once
+	value    T
+	err      error
+}
+
+func (l *LazyInit[T]) Get() (T, error) {
+	l.initOnce.Do(func() {
+		l.value, l.err = l.init()
+	})
+	return l.value, l.err
+}
+
```

Comment:
I don't see a need to do this in a generic way. I doubt this code will be reused and if it will its small and simple enough that copying it to where its needed is easier than having it in some new package to be re-used 🤷 

Making it implement `io.Closer` also becomes easier if its specific rather than generic:

```go
type LazyOracle struct {
	init  func() (*oracle.WorkOracle, error)
	once  sync.Once
	value *oracle.WorkOracle
	err   error
}

func (l *LazyOracle) Get() (T, error) {
	l.once.Do(func() {
		l.value, l.err = l.init()
	})
	return l.value, l.err
}

func (l *lazyOracle) Close() (error) {
	if l.value != nil {
		return l.value.Close()
	}
}
```